teps/0094-configuring-resources-at-runtime.md

```diff
@@ -52,12 +61,472 @@ While this may be a valuable feature, it should be considered in a separate prop
+```
+
+Example `TaskRun`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  name: image-build-taskrun
+spec:
+  taskRef:
+    name: image-build-task
+  stepOverrides:
+    - taskRunStepName: build
+      resources:
+        requests:
+          memory: 1Gi
+```
+
+### Example Pipeline and PipelineRun
+
+Example `Pipeline`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: image-build-pipeline
+spec:
+  tasks:
+    - name: image-build-task
+      steps:
+       - name: build
+         image: gcr.io/kaniko-project/executor:latest
+         command:
+           - /kaniko/executor
+```
+
+Example `PipelineRun`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: image-build-pipelinerun
+spec:
+  taskRunSpecs:
+    - pipelineTaskName: image-build-task
+      stepOverrides:
+        - taskRunStepName: build
+          resources:
+            requests:
+              memory: 1Gi
+```
+
+### Notes/Caveats
+
+#### Mapping container overrides to Steps and Sidecars
+
+This TEP proposes mapping container overrides to their corresponding `Step`s and `Sidecar`s
+using named subobjects, as recommended by
+Kubernetes [API convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps).
+This strategy is consistent with other parts of the Tekton API,
+such as the use of
+[`PipelineTaskRunSpec`](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#specifying-taskrunspecs)
+to specify `TaskRun` configuration for a `Task` in a `Pipeline`.
+It also meets the requirement that users can specify resource requirements for individual `Step`s and `Sidecar`s.
+
+`Step`s and `Sidecar`s are treated separately because they have different fields.
+We may at some point want to override `Step` fields that are not present in `Sidecar`,
+or vice versa. In addition, a `Step` could share a name with a `Sidecar`;
+separating `StepOverrides` and `SidecarOverrides` avoids ambiguity in this case.
+Names that don't match `Step` or `Sidecar` names will result in the `TaskRun` being rejected.
+
+Some users may want to have resource requirements apply to every `Step` or `Sidecar`.
+This TEP proposes that `TaskRunStepOverride`s without `TaskRunStepName`s specified will apply
+to all `Step`s in the corresponding `Task`, and likewise for `Sidecar`s.
+If such a `TaskRunStepOverride` is present, it must be the only entry in `StepOverrides`.
+This will be the only way to override resource requirements of unnamed `Step`s or `Sidecar`s.
+
+An alternative option is to use a map of `Step` or `Sidecar` names to container overrides,
+but this violates Kubernetes API convention.
+We could also apply container overrides to the `Step` or `Sidecar` with the same index.
+This strategy could result in unintended overrides in a `TaskRun`
+if `Step`s are reordered, added, or removed in the corresponding `Task`.
```

Comment:
True - if a step is renamed, an override will stop working, but if new steps are added it will not be obvious that the overrides is hitting the wrong step. The order of steps is not what we consider as the external stable API of a task, so the step index might change for a task in the catalog with no change in version number.

Reply:
I agree! (Helpful to hear that you have given thought to what is the stable API for a task-- any more resources you can point me to?) It doesn't sound like you are requesting any changes here but I want to double check.

Reply:
My comment was based on the criteria we follow in the catalog to decide when changes to a task require a new task version: https://github.com/tektoncd/community/blob/main/teps/0003-tekton-catalog-organization.md#versioning-resources